src/contracts/Executor.sol

```diff
@@ -0,0 +1,121 @@
+// The Licensed Work is (c) 2022 Sygma
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.11;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/Context.sol";
+
+
+import "./interfaces/IBridge.sol";
+import "./interfaces/IHandler.sol";
+
+
+/**
+    @title Facilitates proposal executions.
+    @author ChainSafe Systems.
+ */
+contract Executor is Context {
+    using ECDSA for bytes32;
+
+    IBridge public immutable _bridge;
+
+    // origin domainID => nonces set => used deposit nonces
+    mapping(uint8 => mapping(uint256 => uint256)) public usedNonces;
+
+    struct Proposal {
+        uint8 originDomainID;
+        uint64 depositNonce;
+        bytes32 resourceID;
+        bytes data;
+    }
+
+    event ProposalExecution(uint8 originDomainID, uint64 depositNonce, bytes32 dataHash, bytes handlerResponse);
+    event FailedHandlerExecution(bytes lowLevelData, uint8 originDomainID, uint64 depositNonce);
+
+    error EmptyProposalsArray();
+    error BridgeIsPaused();
+
+    modifier whenBridgeNotPaused() {
+        if (_bridge.paused()) revert BridgeIsPaused();
+        _;
+    }
+
+    constructor(address bridge) {
+        _bridge = IBridge(bridge);
+    }
+
+
+    /**
+        @notice Executes a batch of deposit proposals using a specified handler contract for each proposal
+        @notice If executeProposals fails it doesn't revert, emits {FailedHandlerExecution} event.
+        @param proposals Array of Proposal which consists of:
+        - originDomainID ID of chain deposit originated from.
+        - resourceID ResourceID to be used when making deposits.
+        - depositNonce ID of deposit generated by origin Bridge contract.
+        - data Data originally provided when deposit was made.
+        @notice Emits {ProposalExecution} event for each proposal in the batch.
+        @notice Behaviour: when execution fails, the handler will terminate the function with revert.
+     */
+    function executeProposals(Proposal[] memory proposals) public whenBridgeNotPaused {
```

Comment:
Access control? There should be a test for this.

Reply:
I mean, executeProposal can be called by anyone, which does not seem correct.

Reply:
Yes, yes, should've be more descriptive about this. It looks a bit weird right now, this is a middle step and in the following PR I'll add proof verification logic to executeProposal, where on source chain we'll store a hash of all the data in the deposit and on the destination chain we'll check if that has been included in the block state.